src/librustc_typeck/outlives/implicit_infer.rs

```diff
@@ -0,0 +1,257 @@
+
+    fn visit_trait_item(&mut self, trait_item: &'tcx hir::TraitItem) {}
+
+    fn visit_impl_item(&mut self, impl_item: &'tcx hir::ImplItem) {}
+}
+
+fn required_predicates_to_be_wf<'tcx>(
+    tcx: TyCtxt<'_, 'tcx, 'tcx>,
+    field_def: &ty::FieldDef,
+    infered_outlives_map: &FxHashMap<DefId, Rc<Vec<ty::Predicate<'tcx>>>>,
+) -> HashSet<ty::Predicate<'tcx>> {
+
+    let mut predicates = HashSet::new();
+    // Get the type of the field with identity substs applied.  For
+    // now, let's just see if that causes horrible cycles.
+    let field_ty = tcx.type_of(field_def.did);
+    match field_ty.sty {
+        // For each type `&'a T`, we require `T: 'a`
+        //
+        // If you have A = &'a T, then mt.ty here represents T and
+        // ty represents A (and r represents 'a). So we want T: 'a.
+        ty::TyRef(r, mt) => {
+            let outlives_pred = ty::Binder(ty::OutlivesPredicate(mt.ty, r))
+                .to_predicate();
+            predicates.insert(outlives_pred);
+        }
+
+        // For each struct/enum/union type `Foo<'a, T>`, we can
+        // load the current set of inferred and explicit predicates from
+        // `inferred_outlives_map` and see if those include `T: 'a`
+        ty::TyAdt(def, substs) => {
+
+            // Get type generics of def and lifetime(region) generics of def
+            let generics: &ty::Generics = tcx.generics_of(def.did);
+            // let gen_types: &Vec<ty::TypeParameterDef> = &generics.types;
+            // let gen_region: &Vec<ty::RegionParameterDef> = &generics.regions;
+            let gen_types: &HashSet<DefId> = &generics.types
+                .iter().map(|t| t.def_id).collect();
+            let gen_region: &HashSet<DefId> = &generics.regions
+                .iter().map(|t| t.def_id).collect();
+
+            // Iterate over all predicates in the infered_outlives_map.
+            // See if there are any OutlivesPredicate kind. If there
+            // exists a type and region that is also in this field
+            // then that should be returned.
+            for (_, p_list) in infered_outlives_map.iter() {
+                for p in p_list.as_ref() {
+                    match p {
+                        // ty::Predicate::Trait(..) |
+                        // ty::Predicate::Equate(..) |
+                        // ty::Predicate::Subtype(..) |
+                        // ty::Predicate::Projection(..) |
+                        // ty::Predicate::ClosureKind(..) |
+                        // ty::Predicate::ObjectSafe(..) |
+                        // ty::Predicate::ConstEvaluatable(..) |
+                        // ty::Predicate::WellFormed(subty) => |
+                        // ty::Predicate::RegionOutlives(ref data) => 5,
+
+                        ty::Predicate::TypeOutlives(ty::Binder(
+                            ty::OutlivesPredicate(&typ, reg)
+                        )) => {
+
+                            // FIXME get the DefId from the Ty<'tcx> field
```

Comment:
@nikomatsakis is there a helper function I can use to get the DefId or should I implementing this differently? From asking around rustc not all Ty<> had a DefId